Generate scalafmt conf file on fmt command (#1192)
VirtusLab/scala-cli: website/docs/commands/fmt.md

scala-cli fmt --check
```

### Scalafmt version and dialect

> Configuration file with specified **version** and **dialect** is required to run `scalafmt`. Scala CLI `fmt` tries to find the `.scalafmt.conf` file in the **projects root** directory and **current workspace** directory. There are 3 possible cases:

Reviewer: Actually, the configuration file is no longer required, the way I understand it.
Instead:
```suggestion
Specifying the  **version** and **dialect** is required to run `scalafmt`. Those can be passed through the `scalafmt` configuration file or directly with Scala CLI. If none is specified explicitly, default values depending on the context will be used.

Scala CLI's `fmt` sub-command tries to find the `.scalafmt.conf` file in the **projects root** directory and **current workspace** directory. 

There are 3 possible cases:
```

Author: Scala CLI `fmt` command runs `scalafmt` under the hood and as far as I know `scalafmt` still needs `.scalafmt.conf` file.

Author: Maybe I should start with information that `fmt` command runs `scalafmt` under the hood, but this is how description of [fmt](https://scala-cli.virtuslab.org/docs/commands/fmt) starts.

Reviewer: we possibly could also clarify what do we mean by the project root here (since here we calculate it based on the existing `git` configuration, and not based on the directory of the first input, like in other sub-commands)

Reviewer: > Scala CLI fmt command runs scalafmt under the hood and as far as I know scalafmt still needs .scalafmt.conf file.

Ok, fair point.

What I mean is - from the user's perspective, the important part is the `fmt` sub-command, which now does not need the conf file. 

You're right that we can and should clarify how we do it under the hood (since one can peek into the `.scala-build` folder on how we do it), but IMO the reader of this doc should be clear that when using the `fmt` command, he doesn't need to specify anything, really (as long as he's fine with the defaults we infer)

Author: Sounds good. I will take care of it.

1. Configuration file with the specified version and dialect is found.

> 2. Configuration file is found, but it doesn't have specified version or dialect.

Reviewer: NIT: technically, you missed the few extra cases when one is found, but not the other

Author: Isn't subsection 2 covering it?

Reviewer: it is, just pointing out there's actually more than 3 cases, which create a slight dissonance here

3. Configuration file is not found.

- In the **first** case `fmt` uses the found `.scalafmt.conf` file to run `scalafmt`.
- In the **second** case `fmt` creates a `.scalafmt.conf` file inside the `.scala-build` directory. Content of the previously found file is copied into the newly created file and missing parameters are written into it. New file is used to run `scalafmt`. 
- In the **third** case `fmt` creates a `.scalafmt.conf` file inside the `.scala-build` directory, writes a default version and a dialect into it and uses it to run `scalafmt`.

If the `--save-scalafmt-conf` option is passed, then `fmt` command behaves as follows:
- In the **first** case `fmt` uses the found `.scalafmt.conf` file to run `scalafmt`.
- In the **second** case `fmt` writes missing parameters directly into the previously found file and then uses it to run `scalafmt`.
- In the **third** case `fmt` creates a `.scalafmt.conf` file in the current workspace directory, writes a default version and a dialect into it and uses it to run `scalafmt`.

> It is also possible to pass `--dialect` option to define which dialect should be written into configuration file, which will be used to run `scalafmt` command:

Reviewer: IMO, we should also enable passing `--scalafmt-version` in a similar fashion from the command line

Reviewer: whatever is passed should also take priority over the `.scalafmt.conf` contents, which IMO could be clarified further here. (I think it's already the case for `--dialect`)

scala-cli fmt --check
```

### Scalafmt version and dialect

> Scala CLI `fmt` command runs `scalafmt` under the hood, which requires `.scalafmt.conf` configuration file with specified **version** and **dialect**. 

Reviewer: I meant something more along the lines of:
```suggestion
Scala CLI `fmt` command supports passing the dialect and `scalafmt` version directly from the command line, using respectively the `--scalafmt-dialect` and `--scalafmt-version` options. You can even skip passing either of those, which will make Scala CLI to infer a default value. 

If a `.scalafmt.conf` file is present in the workspace, the values will be read from there, unless specified with Scala CLI options.

Otherwise, the default dialect will be inferred based on the configured Scala version (through a Scala CLI option or a using directive), while the default `scalafmt` version will be the latest one used by your Scala CLI version (so it is subject to change when updating Scala CLI).

The Scala CLI `fmt` command runs `scalafmt` under the hood, which *normally* requires `.scalafmt.conf` configuration file with explicitly specified **version** and **dialect** fields. 
The way it is handled by Scala CLI is as follows:
```

scala-cli fmt --check
```

### Scalafmt version and dialect

Scala CLI `fmt` command supports passing the **dialect** and `scalafmt` **version** directly from the command line, using respectively the `--scalafmt-dialect` and `--scalafmt-version` options:
```
scala-cli fmt --scalafmt-dialect scala3 --scalafmt-version 3.5.8
```
You can skip passing either of those, which will make Scala CLI to infer a default value:

> - If a `.scalafmt.conf` file is present in the workspace and it has field defined, the value will be read from there, unless specified with Scala CLI options.

Reviewer: ```suggestion
- If a `.scalafmt.conf` file is present in the workspace and it has the field defined, the value will be read from there, unless explicitly specified with Scala CLI options.
```

- Otherwise, the default `scalafmt` **version** will be the latest one used by your Scala CLI version (so it is subject to change when updating Scala CLI) and the default **dialect** will be inferred based on the previously evaluated `scalafmt` version.

#### Example 1

``` text title=.scalafmt.conf
version = "3.5.8"
runner.dialect = scala212
```

```bash
scala-cli fmt --scalafmt-dialect scala213
```

For above setup `fmt` will use:
- `version="3.5.8"` from the file
- `dialect=scala213`, because passed `--scalafmt-dialect` option overrides dialect found in the file

#### Example 2

``` text title=.scalafmt.conf
version = "2.7.5"
```

```bash
scala-cli fmt --scalafmt-version 3.5.8
```

For above setup `fmt` will use:
- `version="3.5.8"`, because passed `--scalafmt-version` option overrides version from the file

> - `dialect=scala3`, because dialect is neither passed as an option nor is it present in the configuration file, so it is inferred based on the previously evaluated version (`3.5.8`)

Reviewer: uh, the dialect is inferred based on the Scala version, not the `scalafmt` version, no?
since `-S` wasn't passed, the default would be the newest release of dotty: `3.1.3`, as of writing this comment.

So it should be something more like:
```suggestion
- `dialect=scala3`, because dialect is neither passed as an option nor is it present in the configuration file, so it is inferred based on the Scala version; the Scala version wasn't explicitly specified in the command either, so it falls back to the default Scala version - the latest one, thus the resulting dialect is `scala3`. 
```

Author: Oh, thats right. My bad...
